### PR TITLE
Avoid captures in SqlConnection.ReadBlob

### PR DESCRIPTION
**Pull request by sharwell**
This change produced a 806MB (11.3%) allocation reduction in the scenario described by #36114.

### CODE REVIEW

**Comment by Therzok** [comment]
Aren't these changes mostly in https://github.com/dotnet/roslyn/pull/34790?

**Comment by sharwell** [comment]
@Therzok Yes, but I resubmitted them with finer granularity and data against the standardized test

**Comment by sharwell** [comment]
@Therzok It wasn't my intent to duplicate your change. More a side effect of trying to document the baseline values and validate the incremental improvements matched the claims. In the end, it certainly appears that the new test is an accurate representation of the scenario you wanted to improve, since we arrived at very similar conclusions.

**Comment by Therzok** [comment]
Oh, no worries, I was just thinking the commits would be cherry-pickable, but after looking at my PR again,  they touch multiple things in the 'address PR feedback' commits. So looks good to me!

File: src/Workspaces/Core/Portable/Storage/SQLite/Interop/SqlConnection.cs [hunk 1]
```diff
@@ -129,7 +129,18 @@ public ResettableSqlStatement GetResettableStatement(string query)
             return new ResettableSqlStatement(statement);
         }
 
-        public void RunInTransaction(Action action)
+        public void RunInTransaction<TState>(Action<TState> action, TState state)
+        {
+            RunInTransaction(
+                state =>
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Storage/SQLite/Interop/SqlConnection.cs]: fwiw, calling this state was super confusing.

File: src/Workspaces/Core/Portable/Storage/SQLite/SQLitePersistentStorage_StringIds.cs [hunk 2]
```diff
@@ -86,10 +86,9 @@ private bool TryFetchStringTable(SqlConnection connection)
             // values.
             try
             {
-                connection.RunInTransaction(() =>
-                {
-                    stringId = InsertStringIntoDatabase_MustRunInTransaction(connection, value);
-                });
+                stringId = connection.RunInTransaction(
+                    state => InsertStringIntoDatabase_MustRunInTransaction(state.connection, state.value),
+                    (connection, value));
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Storage/SQLite/SQLitePersistentStorage_StringIds.cs]: nice.

File: src/Workspaces/Core/Portable/Storage/SQLite/Interop/SqlConnection.cs [hunk 3]
```diff
@@ -129,7 +129,18 @@ public ResettableSqlStatement GetResettableStatement(string query)
...
+                {
+                    state.action(state.state);
+                    return (object)null;
```
Comment (Therzok) [inline src/Workspaces/Core/Portable/Storage/SQLite/Interop/SqlConnection.cs]: Maybe? 
```suggestion
                    return default(object);
```

Comment (sharwell) [reply src/Workspaces/Core/Portable/Storage/SQLite/Interop/SqlConnection.cs]: When I see `default(object)`, my first question is "why isn't this just `null`"? When I see `(object)null`, my first thought is "the cast is required for the type system".

File: src/Workspaces/Core/Portable/Storage/SQLite/Interop/SqlConnection.cs (hunk 1 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Storage/SQLite/Interop/SqlConnection.cs]: 😕 I thought that was the normal name for this. Examples of alternatives?

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Storage/SQLite/Interop/SqlConnection.cs]: it's more that the outer param for RunInTransaction is called "state" and hte inner one is called "state" and my brain exploded with confusion.  Honestly, i'd be 100% fine with you calling the inner one `tuple` :)

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Storage/SQLite/Interop/SqlConnection.cs]: it's also a nit.  if you're ok with this, it's NBD.  just giving you my stream of consciousness thoughts as i tried to grok things :)

Comment (drewnoakes) [reply src/Workspaces/Core/Portable/Storage/SQLite/Interop/SqlConnection.cs]: I've seen other instances of this pattern with `TArg` and `arg`.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Storage/SQLite/Interop/SqlConnection.cs]: Note: the pattern is fine.  it's literally the fact that there are two params called 'state', and then the code has `state.state` :D  It's purely naming i don't like.  `arg` or `tuple` would be great IMO just for some simple clarity.

**Pull request merged**